File: app/src/types/challenge.ts
export type ParticipantRole = 'challenger' | 'opponent'

export type CreateChallengeBody = {
  challengerName: string
  opponentName: string
  message?: string
  challengerEmail?: string
  opponentEmail?: string
  /** When set, server verifies via Clerk and may email the opponent */
  challengerClerkUserId?: string
  opponentClerkUserId?: string
  /** When true, deduct 2 credits at send; opponent can accept for free. */
  coverOpponentEntry?: boolean
}

export type Challenge = {
  id: string
  challengerName: string
  opponentName: string
  message?: string
  challengerPushups: number | null
  opponentPushups: number | null
  challengerEmail?: string
  opponentEmail?: string
  challengerClerkUserId?: string
  opponentClerkUserId?: string
  /** Server: proposed | pending | active | completed | declined | cancelled | expired */
  status?: string | null
  /** ISO datetime when the challenge window ends (optional; from API). */
  expiresAt?: string | null
  /** Optional creation time for relative meta (mock or future API). */
  createdAt?: string | null
  /** True when challenger paid 2 credits to cover the opponent's entry. */
  gifted?: boolean
}

export type CreateChallengeResponse = {
  challenge: Challenge
  shareLink: string
  notificationSent: boolean
  /** Set when the API deducts credits (Postgres + billing). */
  balanceAfter?: number | null
}

export type SubmitAttemptBody = {
  participantName: string
  pushupCount: number
  role: ParticipantRole
  /** Optional recording; uploaded only on submit (multipart). */
  video?: Blob
}

export type ChallengeVideoOpponent = {
  username: string
  initials: string
}

export type ChallengeVideoResult = 'won' | 'lost' | 'tie'

export type ChallengeVideoItem = {
  id: string
  challengeId: string
  opponent: ChallengeVideoOpponent
  role: ParticipantRole
  result: ChallengeVideoResult
  yourScore: number
  theirScore: number
  recordedAt: string
  expiresAt: string
  videoUrl: string | null
}

export type ChallengeVideoStats = {
  total: number
  wins: number
  losses: number
  bestReps: number
}

export type ChallengeVideosPage = {
  challengeVideos: ChallengeVideoItem[]
  stats: ChallengeVideoStats
}

export type ChallengeOutcome = {
  challengeId: string
  challengerName: string
  opponentName: string
  challengerPushups: number
  opponentPushups: number
  winner: ParticipantRole | 'tie'
}

export type LeaderboardEntry = {
  rank: number
  displayName: string
  bestPushups: number
  wins: number
  ties: number
  losses: number
}

File: backend/app/db/models/challenge.py
"""Challenge row."""

from __future__ import annotations

from datetime import datetime

from sqlalchemy import Boolean, DateTime, Integer, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base


class Challenge(Base):
    __tablename__ = "challenges"

    id: Mapped[str] = mapped_column(String(36), primary_key=True)
    challenger_name: Mapped[str] = mapped_column(String(255), nullable=False)
    opponent_name: Mapped[str] = mapped_column(String(255), nullable=False)
    message: Mapped[str | None] = mapped_column(Text, nullable=True)
    challenger_pushups: Mapped[int | None] = mapped_column(Integer, nullable=True)
    opponent_pushups: Mapped[int | None] = mapped_column(Integer, nullable=True)
    challenger_email: Mapped[str | None] = mapped_column(String(512), nullable=True)
    opponent_email: Mapped[str | None] = mapped_column(String(512), nullable=True)
    challenger_clerk_user_id: Mapped[str | None] = mapped_column(
        String(255), nullable=True, index=True
    )
    opponent_clerk_user_id: Mapped[str | None] = mapped_column(
        String(255), nullable=True, index=True
    )
    status: Mapped[str] = mapped_column(
        String(32),
        nullable=False,
        server_default="pending",
        index=True,
    )
    initiator_clerk_user_id: Mapped[str | None] = mapped_column(String(255), nullable=True)
    opponent_is_member: Mapped[bool | None] = mapped_column(Boolean, nullable=True)
    rules_json: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
        onupdate=func.now(),
    )
    completed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    expires_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    idempotency_key: Mapped[str | None] = mapped_column(String(128), nullable=True)
    # True = challenger paid for opponent's entry (2 credits at send; opponent pay 0 on accept)
    gifted: Mapped[bool] = mapped_column(
        Boolean, nullable=False, server_default="false"
    )

    attempts: Mapped[list["ChallengeAttempt"]] = relationship(
        "ChallengeAttempt",
        back_populates="challenge",
        cascade="all, delete-orphan",
    )
